src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py

```diff
@@ -1370,6 +1370,43 @@ def aks_upgrade(cmd, client, resource_group_name, name, kubernetes_version, no_w
     return client.create_or_update(resource_group_name, name, instance, raw=no_wait)
 
 
+def _ensure_aks_service_principal(cli_ctx,
+                                  service_principal=None,
+                                  client_secret=None,
+                                  subscription_id=None,
+                                  dns_name_prefix=None,
+                                  location=None,
+                                  name=None):
```

Comment:
if the difference from the existing `_ensure_service_principal` is only on role assignments, why not just expose an argument of `skip_assignment` from the existing one, this way we can eliminate quite a few dupes

Reply:
I don't want to make many switches in the _ensure_service_principal function. 
Because,
1. It will make changes to existing `az acs` command, and we are trying to decouple them. 
2. I would expect more changes in aks, and very little change in acs. It would finally not as similar logic as today.

So, I prefer the current way.

Reply:
I see. Moving forward we probably should just split out the ACS logic in a different file.

Reply:
Yeah, that would be the way going forward.